sparse/impl/KokkosSparse_sptrsv_solve_impl.hpp

```diff
@@ -121,91 +143,532 @@ struct SptrsvWrap {
+                                                                            b);
+    }
+
+    // multiply_subtract. C -= A * B
+    KOKKOS_INLINE_FUNCTION
+    static void multiply_subtract(const CBlock &A, const CVector &B,
+                                  ArrayType &Ca) {
+      Vector C(&Ca.m_data[0], B.size());
+      multiply_subtract(A, B, C);
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    static void multiply_subtract(const CBlock &A, const CVector &B,
+                                  Vector &C) {
+      // Use gemv. alpha is hardcoded to -1, beta hardcoded to 1
+      KokkosBlas::SerialGemv<KokkosBlas::Trans::NoTranspose,
+                             KokkosBlas::Algo::Gemv::Blocked>::invoke(-1.0, A,
+                                                                      B, 1.0,
+                                                                      C);
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    static void copy(const member_type &team, const Vector &lhs_,
+                     ArrayType &rhsa) {
+      CVector rhs_(&rhsa.m_data[0], lhs_.size());
+      assign(team, lhs_, rhs_);
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    static void copy(const Vector &lhs_, ArrayType &rhsa) {
+      CVector rhs_(&rhsa.m_data[0], lhs_.size());
+      assign(lhs_, rhs_);
+    }
+
+    // lget
+    KOKKOS_INLINE_FUNCTION
+    Vector lget(const size_type row) const {
+      return Vector(lhs.data() + (row * block_size), block_size);
+    }
+
+    // rget
+    KOKKOS_INLINE_FUNCTION
+    CVector rget(const size_type row) const {
+      return CVector(rhs.data() + (row * block_size), block_size);
+    }
+
+    // vget
+    KOKKOS_INLINE_FUNCTION
+    CBlock vget(const size_type block) const {
+      return CBlock(values.data() + (block * block_items), block_size,
+                    block_size);
+    }
+
+    // lhs = (lhs + rhs) / diag
+    KOKKOS_INLINE_FUNCTION
+    static void add_and_divide(const member_type &team, const Vector &lhs_val,
+                               const CVector &rhs_val, const CBlock &diag_val) {
+      add(team, rhs_val, lhs_val);
+      divide(team, lhs_val, diag_val);
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    static void add_and_divide(const Vector &lhs_val, const CVector &rhs_val,
+                               const CBlock &diag_val) {
+      add(rhs_val, lhs_val);
+      divide(lhs_val, diag_val);
+    }
+
+    // print
+    KOKKOS_INLINE_FUNCTION
+    static void print(const CBlock &item) {
+      std::cout << "Block: ";
+      for (size_type i = 0; i < item.extent(0); ++i) {
+        std::cout << "      ";
+        for (size_type j = 0; j < item.extent(1); ++j) {
+          std::cout << item(i, j) << " ";
+        }
+        std::cout << std::endl;
+      }
+    }
+
+    // print
+    KOKKOS_INLINE_FUNCTION
+    static void print(const CVector &item) {
+      std::cout << "Vector: ";
+      for (size_type i = 0; i < item.extent(0); ++i) {
+        std::cout << item(i) << " ";
+      }
+      std::cout << std::endl;
+    }
+
```

Comment:
```suggestion
```
Same as above, can't ``std::cout`` in device code

Reply:
Done, removed.